wallet/wallet-controller/src/helpers.rs

```diff
@@ -0,0 +1,298 @@
+            .await
+            .map_err(ControllerError::NodeCallError)?
+    } else {
+        get_tx_output_destination(&utxo, &|_| None, HtlcSpendingCondition::Skip)
+    }
+    .ok_or(ControllerError::WalletError(WalletError::CannotFindUtxo(
+        input.clone(),
+    )))?;
+
+    Ok((utxo, dest))
+}
+
+fn pool_id_from_txo(utxo: &TxOutput) -> Option<PoolId> {
+    match utxo {
+        TxOutput::CreateStakePool(pool_id, _) | TxOutput::ProduceBlockFromStake(_, pool_id) => {
+            Some(*pool_id)
+        }
+        TxOutput::Burn(_)
+        | TxOutput::Transfer(_, _)
+        | TxOutput::LockThenTransfer(_, _, _)
+        | TxOutput::Htlc(_, _)
+        | TxOutput::CreateOrder(_)
+        | TxOutput::IssueNft(_, _, _)
+        | TxOutput::IssueFungibleToken(_)
+        | TxOutput::DelegateStaking(_, _)
+        | TxOutput::CreateDelegationId(_, _)
+        | TxOutput::DataDeposit(_) => None,
+    }
+}
+
+async fn fetch_token_extra_info<T>(
+    rpc_client: &T,
+    value: &OutputValue,
+) -> Result<Option<TokenAdditionalInfo>, ControllerError<T>>
+where
+    T: NodeInterface,
+{
+    match value {
+        OutputValue::Coin(_) | OutputValue::TokenV0(_) => Ok(None),
+        OutputValue::TokenV1(token_id, _) => {
+            let info = fetch_token_info(rpc_client, *token_id).await?;
+            Ok(Some(TokenAdditionalInfo {
+                num_decimals: info.token_number_of_decimals(),
+                ticker: info.token_ticker().to_vec(),
+            }))
+        }
+    }
+}
+
+pub async fn fetch_utxo_extra_info<T>(
+    rpc_client: &T,
+    utxo: TxOutput,
+) -> Result<UtxoWithAdditionalInfo, ControllerError<T>>
+where
+    T: NodeInterface,
+{
+    match &utxo {
+        TxOutput::Burn(value)
+        | TxOutput::Transfer(value, _)
+        | TxOutput::LockThenTransfer(value, _, _)
+        | TxOutput::Htlc(value, _) => {
+            let additional_info = fetch_token_extra_info(rpc_client, value)
+                .await?
+                .map(UtxoAdditionalInfo::TokenInfo);
+            Ok(UtxoWithAdditionalInfo::new(utxo, additional_info))
+        }
+        TxOutput::CreateOrder(order) => {
+            let ask = fetch_token_extra_info(rpc_client, order.ask()).await?;
+            let give = fetch_token_extra_info(rpc_client, order.ask()).await?;
+            let additional_info = Some(UtxoAdditionalInfo::CreateOrder { ask, give });
+            Ok(UtxoWithAdditionalInfo::new(utxo, additional_info))
+        }
+        TxOutput::ProduceBlockFromStake(_, pool_id) => {
+            let staker_balance = rpc_client
+                .get_staker_balance(*pool_id)
+                .await
+                .map_err(ControllerError::NodeCallError)?
+                .ok_or(WalletError::UnknownPoolId(*pool_id))?;
+            Ok(UtxoWithAdditionalInfo::new(
+                utxo,
+                Some(UtxoAdditionalInfo::PoolInfo { staker_balance }),
+            ))
+        }
+        TxOutput::IssueNft(_, _, _)
```

Comment:
I find this somewhat scary. E.g. we're signing a raw transaction, which was created who knows when, and then here we're obtaining the pool balance which now may be different from the one that it had when the tx was created (e.g. due to a reorg). So we're basically producing potentially incorrect data and rely on it being ignored (because, as I can see, it's not used in places where this function is called).
Which makes me think that "dynamic" data, like pool balances (as opposed to "static" data, like token decimals) shouldn't be in UtxoWithAdditionalInfo.
But then it'd make sense to move the static data out of it as well, because e.g. it'd be more natural for token decimals to be stored in a single map and passed alongside `PartiallySignedTransaction` (or maybe it can still be a part of `PartiallySignedTransaction`, but not a part of individual utxos).
I wonder whether we should reconsider this whole approach with `UtxoWithAdditionalInfo`. @azarovh what do you think?

Reply:
Not sure there is a way around it because we have to display what is being signed and a tx has to have an output with specific decommissioned amount.

Reply:
The thing is that the same dynamic data is also used on tx creation and once you submit the tx to the node to validate the tx. So if it is wrong now it will also be wrong on submittion.

The only issue I see is if the user's node itself creates 2 blocks which cause a reorg. They create the tx while on the first block and then a reorg happens with the other block which can have more or less fees so the tx might be rejected if the new balance is less or silently accepted and the excess coins going to fees.